packages/front-end/components/PowerCalculation/stats.ts

```diff
@@ -340,3 +345,543 @@ export function powerMetricWeeks(
+    priorVarianceRelDGP,
+    priorMeanRelSpecified,
+    priorVarianceRelSpecified,
+    proper,
+    mean,
+    variance,
+    nPerVariation,
+    relative
+  );
+  while (currentPower > power) {
+    lowerBound *= 0.5;
+    currentPower = powerEstBayesian(
+      alpha,
+      lowerBound,
+      priorVarianceRelDGP,
+      priorMeanRelSpecified,
+      priorVarianceRelSpecified,
+      proper,
+      mean,
+      variance,
+      nPerVariation,
+      relative
+    );
+  }
+  /*maximum solution is 200% increase*/
+  const maxPower = powerEstBayesian(
+    alpha,
+    2,
+    priorVarianceRelDGP,
+    priorMeanRelSpecified,
+    priorVarianceRelSpecified,
+    proper,
+    mean,
+    variance,
+    nPerVariation,
+    true
+  );
+  if (maxPower < power) {
+    console.log(`failing at iteration j: %d`, j);
+    console.log(`maxPower: %d`, maxPower);
+    console.log(`priorVariancenRelDGP: %d`, priorVarianceRelDGP);
+    console.log(`priorMeanRelSpecified: %d`, priorMeanRelSpecified);
+    console.log(`priorVarianceRelSpecified: %d`, priorVarianceRelSpecified);
+    console.log(
+      `mean: %d, variance: %d, nPerVariation %d`,
+      mean,
+      variance,
+      nPerVariation
+    );
+    const mdeResults: MDEResults = {
+      type: "error",
+      description:
+        "need to increase number of users or reduce number of variations.",
+    };
+    return mdeResults;
+  }
+  let upperBound = 2 * lowerBound;
+  currentPower = powerEstBayesian(
+    alpha,
+    upperBound,
+    priorVarianceRelDGP,
+    priorMeanRelSpecified,
+    priorVarianceRelSpecified,
+    proper,
+    mean,
+    variance,
+    nPerVariation,
+    relative
+  );
+  while (currentPower < power) {
+    upperBound *= 2;
+    currentPower = powerEstBayesian(
+      alpha,
+      upperBound,
+      priorVarianceRelDGP,
+      priorMeanRelSpecified,
+      priorVarianceRelSpecified,
+      proper,
+      mean,
+      variance,
+      nPerVariation,
+      relative
+    );
+  }
+  let mde = 0.5 * (lowerBound + upperBound);
+  currentPower = powerEstBayesian(
+    alpha,
+    mde,
+    priorVarianceRelDGP,
+    priorMeanRelSpecified,
+    priorVarianceRelSpecified,
+    proper,
+    mean,
+    variance,
+    nPerVariation,
+    relative
+  );
+  let diff = currentPower - power;
+  let iters = 0;
+  while (Math.abs(diff) >= 1e-5 && iters < 1e5) {
+    if (diff > 0) {
+      upperBound = mde;
+    } else {
+      lowerBound = mde;
+    }
+    mde = 0.5 * (lowerBound + upperBound);
+    currentPower = powerEstBayesian(
+      alpha,
+      mde,
+      priorVarianceRelDGP,
+      priorMeanRelSpecified,
+      priorVarianceRelSpecified,
+      proper,
+      mean,
+      variance,
+      nPerVariation,
+      relative
+    );
+    diff = currentPower - power;
+    iters += 1;
+  }
+  /*case where mde converged to nonzero value*/
+  const mdeResults: MDEResults = {
+    type: "success",
+    mde: mde,
+  };
+  return mdeResults;
+}
+
+export function powerMetricWeeksBayesian(
```

Comment:
I see why you asked about copying. There's a lot of complex logic here that it might benefit us to only write once. Take a look at this slight rewrite https://github.com/growthbook/growthbook/compare/ls/power-structure?expand=1

I've done two key things:
* Make some of the code a bit more typescript friendly (e.g. don't unpack a bunch of stuff, use .forEach)
* Ask you to change `powerEst` and `powerEstBayesian` to instead take an object with your parameters specified. Then build those parameters and specify which function to use in one single if-else block.

My code is psuedo-code, since I didn't create those other types, and can just serve as a reference.

Reply:
I incorporated your pseudo-code.  However, I didn't define new types as inputs for `powerEst` and `powerEstBayesian`, as the linter was fussing at me for having different types as inputs for a `powerFunction` function.  Also, the way I have it is simple.  However, if you have more elegant approach, I'm open to incorporating it.  Also, I have `statsEngineSettings` as attribute of `powerSettings`, rather than standalone object, because we always want the freq statsEngineSettings to be part of `PowerCalculationParams` and bayesian statsEngineSettings to be associated with PowerCalculationParamsBayesian.  Again, if you have better suggestion, I'm open.